### ISSUE TITLE
Why do queries optimized with projection read less rows, but consume more memory and take longer to execute?

### ISSUE THREAD

**Issue by fsgdgsdfgsgd** [ClickHouse/ClickHouse#53509]
Hi ClickHouse devs,

We recently did some tests on projection feature, but I encountered a special case and I want to know the reason. Because some security risk, I hid the database name, table name and column name. But the basics of the query are the same.
The projection definition was like this:
```
ALTER TABLE {database}.{table} ON CLUSTER {cluster_name} ADD PROJECTION test_projection_1 (
    SELECT
        toStartOfHour(toDateTime(processDate)),
        count(DISTINCT userId)
    GROUP BY processDate, condition_1, condition_2, condition_3, toStartOfHour(toDateTime(processDate))
);
```

The test query was like this:
```
SELECT
    toStartOfHour(toDateTime(processDate)) AS __timestamp,
    count(DISTINCT userId) AS "USERID"
FROM {database}.{table}
WHERE processDate >= toDateTime('2023-08-09')
    AND processDate < toDateTime('2023-08-16')
    AND (((condition_1 = '1'
                OR condition_2 = '1'
                OR condition_3 = '1')))
GROUP BY toStartOfHour(toDateTime(processDate))
settings optimize_use_projections=1;
```
We used 'settings' to control if it should use projection.

The query result with the projection optimization
```
Memoery Usage: 4.35GB
Query Result: 169 rows in set. Elapsed: 15.768 sec. Processed 24.29 million rows, 5.42 GB (1.54 million rows/s., 343.56 MB/s.)
```

The query result without the projection optimization
```
Memoery Usage: 0.34GB
Query Result: 169 rows in set. Elapsed: 3.913 sec. Processed 27.43 billion rows, 1.28 TB (7.01 billion rows/s., 327.02 GB/s.)
```

We can notice that the query that used the projection read less rows, but consumed more memory and took longer to finish. This is a little bit wired because in general if a query reads less rows, it should also consume less memory and execute faster. Does anyone know the reason for this?

**Comment by UnamedRus** [comment]
https://github.com/ClickHouse/ClickHouse/issues/53057#issuecomment-1666201003

Check this explanation, it's about AggregatingMergeTree, but the same applies for Projections.
TLDR, your aggregated projection row on average much heavier than row from raw table.
Plus default index granularity 8192, does mean that ClickHouse will read more data than it's needed.

>        count(DISTINCT userId)

Don't use uniqExact functions in projections/aggregatingMergeTree tables.
Use uniqCombined instead, their state much smaller
https://kb.altinity.com/altinity-kb-schema-design/uniq-functions/#functions-to-count-uniqs

**Comment by den-crane** [comment]
https://stackoverflow.com/questions/76592036/clickhouse-materialized-view-on-replicatedaggregatingmergetree/76592492#76592492
